File: retort-core/src/main/java/co/kaioru/retort/builder/ICommandBuilder.java
package co.kaioru.retort.builder;

import co.kaioru.retort.ICommand;
import co.kaioru.retort.ICommandContext;
import co.kaioru.retort.ICommandExecutable;
import co.kaioru.retort.ICommandMiddleware;

public interface ICommandBuilder<I extends ICommandContext, O> {

	ICommandBuilder withDescription(String description);

	ICommandBuilder withAlias(String alias);

	ICommandBuilder withMiddleware(ICommandMiddleware<I> middleware);

	ICommandBuilder withCommand(ICommand<I, O> command);

	ICommand<I, O> build(ICommandExecutable<I, O> executable);

}

File: retort-core/src/main/kotlin/co/kaioru/retort/builder/ICommandBuilder.kt
package co.kaioru.retort.builder

import co.kaioru.retort.ICommand
import co.kaioru.retort.ICommandExecutable
import co.kaioru.retort.ICommandMiddleware

interface ICommandBuilder<I, O> : ICommand<I, O> {
    fun withDescription(description: String): ICommandBuilder<I, O> {
        this.description = description
        return this
    }

    fun withAlias(alias: String): ICommandBuilder<I, O> {
        registerAlias(alias)
        return this
    }

    fun withAliases(aliases: Collection<String>): ICommandBuilder<I, O> {
        registerAliases(aliases)
        return this
    }

    fun withMiddleware(middleware: ICommandMiddleware<I>): ICommandBuilder<I, O> {
        registerMiddleware(middleware)
        return this
    }

    fun withMiddleware(middleware: Collection<ICommandMiddleware<I>>): ICommandBuilder<I, O> {
        registerMiddleware(middleware)
        return this
    }

    fun withCommand(command: ICommand<I, O>): ICommandBuilder<I, O> {
        registerCommand(command)
        return this
    }

    fun withCommands(commands: Collection<ICommand<I, O>>): ICommandBuilder<I, O> {
        registerCommands(commands)
        return this
    }

    fun build(executable: ICommandExecutable<I, O>): ICommand<I, O>
}